File: web/lib/nl.ts
/**
 * Natural-language decision engine (TypeScript, runs inside Next.js).
 *
 * Same law as the Python version: retrieve structured evidence FIRST, then let
 * the AI reason over it and cite figures. With no AI provider it answers
 * deterministically from the retrieved evidence. Reads through the query layer,
 * so it works against Postgres or the static snapshot.
 */

import "server-only";
import { getKpis, getRanking, getForecast, getRecommendations } from "./queries";
import { generate, isConfigured, providerLabel } from "./ai/provider";
import { PERSONAS, PersonaKey } from "./ai/personas";

export type NlAnswer = {
  question: string;
  answer: string;
  citations: string[];
  evidence: Record<string, unknown>;
  provenance: { engine: "deterministic" | "ai-enhanced"; provider: string; persona: string };
};

const has = (q: string, words: string[]) => words.some((w) => q.includes(w));

async function retrieve(q: string) {
  const lower = q.toLowerCase();
  const kpis = await getKpis();
  const rate = kpis.find((k) => k.metric_key === "placement_rate");
  const ctc = kpis.find((k) => k.metric_key === "median_ctc");

  const evidence: Record<string, unknown> = {
    headline: {
      placement_rate: rate?.value, placement_rate_delta: rate?.delta,
      median_ctc: ctc?.value, median_ctc_delta: ctc?.delta,
    },
  };
  const citations: string[] = [];
  if (rate) citations.push(`Placement rate ${rate.value.toFixed(1)}% (${(rate.delta ?? 0) >= 0 ? "+" : ""}${(rate.delta ?? 0).toFixed(1)} pts vs prior cycle).`);
  if (ctc) citations.push(`Median CTC ₹${ctc.value.toFixed(1)} LPA (${(ctc.delta ?? 0) >= 0 ? "+" : ""}${(ctc.delta ?? 0).toFixed(1)} LPA vs prior cycle).`);

  if (has(lower, ["region", "underperform", "where", "geograph"])) {
    const region = await getRanking("region_placement_rate");
    if (region.length) {
      evidence.regions = region;
      citations.push(`${region[0].entity} is the weakest region at ${region[0].metric.toFixed(1)}% placement.`);
    }
  }
  if (has(lower, ["skill", "premium", "critical", "demand", "pay"])) {
    const prem = await getRanking("skill_salary_premium");
    if (prem.length) {
      evidence.skill_premium = prem.slice(0, 5);
      citations.push(`Top skill premium: ${prem[0].entity} at ${prem[0].metric >= 0 ? "+" : ""}${prem[0].metric.toFixed(1)}% over median.`);
    }
  }
  if (has(lower, ["channel", "source", "hire", "recruit"])) {
    const ch = await getRanking("channel_effectiveness");
    if (ch.length) {
      evidence.channels = ch;
      citations.push(`${ch[0].entity} is the highest-yield channel at ${ch[0].metric.toFixed(1)}%.`);
    }
  }
  if (has(lower, ["ctc", "salary", "forecast", "next", "future", "project"])) {
    const fc = await getForecast("median_ctc_cycle");
    const proj = fc.find((p) => p.is_forecast);
    if (proj) {
      evidence.ctc_forecast_next = proj.yhat;
      citations.push(`Median CTC projected to ₹${proj.yhat.toFixed(1)} LPA next cycle.`);
    }
  }
  if (has(lower, ["sector", "revenue", "decline", "fall", "fell", "drop", "held back", "it services"])) {
    const recs = await getRecommendations();
    const it = recs.find((r) => /IT Services/.test(r.title) || /IT Services/.test(r.observation));
    if (it) {
      evidence.declining_sector = { title: it.title, observation: it.observation };
      citations.push(it.observation);
    }
  }
  if (has(lower, ["invest", "budget", "allocate", "should we", "priorit"])) {
    const recs = await getRecommendations();
    evidence.recommendations = recs.slice(0, 4).map((r) => ({
      title: r.title, action: r.recommended_action, priority: r.priority,
    }));
  }

  return { evidence, citations };
}

export async function answer(question: string, persona: PersonaKey = "executive"): Promise<NlAnswer> {
  const { evidence, citations } = await retrieve(question);

  if (isConfigured()) {
    const prompt =
      `Question: "${question}"\n\n` +
      "Answer the question using ONLY the evidence below. Cite the figures you use. " +
      "If the evidence does not answer it, say so plainly. " +
      "Keep it tight: one short paragraph, then if there are actions put each on its own " +
      "line starting with '- '. Use **bold** only for a few key terms.\n\nEVIDENCE:\n" +
      JSON.stringify(evidence, null, 2);
    const text = await generate({ system: PERSONAS[persona].system, prompt, temperature: 0.3 });
    if (text) {
      return {
        question, answer: text, citations, evidence,
        provenance: { engine: "ai-enhanced", provider: providerLabel(), persona },
      };
    }
  }

  const body = citations.length ? citations.join(" ") : "No structured evidence matched this question.";
  return {
    question, answer: `Based on the placement analytics: ${body}`, citations, evidence,
    provenance: { engine: "deterministic", provider: providerLabel(), persona },
  };
}

File: _archive/api/athena_api/nl.py
"""Natural-language decision interface (Layer 5).

The order is the law: retrieve structured analytical results FIRST, then let the
AI reason over them. The question never reaches a model without evidence, and
answers cite the figures they rest on. With no AI provider, a deterministic
answer is composed from the same retrieved evidence.
"""

from __future__ import annotations

from athena.config import MODULE

from . import ai
from .data import forecast, kpis, ranking, recommendations


def _kpi(k, key):
    return next((r for r in k if r["metric_key"] == key), None)


def retrieve(conn, question: str, module=MODULE) -> dict:
    """Pull the evidence relevant to the question. Always includes headline KPIs."""
    q = question.lower()
    k = kpis(conn, module)
    rate, ctc = _kpi(k, "placement_rate"), _kpi(k, "median_ctc")
    evidence: dict = {"headline": {
        "placement_rate": rate and rate["value"], "placement_rate_delta": rate and rate["delta"],
        "median_ctc": ctc and ctc["value"], "median_ctc_delta": ctc and ctc["delta"],
    }}
    citations: list[str] = []

    def cite(text):
        citations.append(text)

    if rate:
        cite(f"Placement rate {rate['value']:.1f}% ({rate['delta']:+.1f} pts vs prior cycle).")
    if ctc:
        cite(f"Median CTC ₹{ctc['value']:.1f} LPA ({ctc['delta']:+.1f} LPA vs prior cycle).")

    if any(w in q for w in ("region", "underperform", "where", "geograph")):
        region = ranking(conn, "region_placement_rate", module)
        if region:
            evidence["regions"] = region
            worst = region[0]
            cite(f"{worst['entity']} is the weakest region at {worst['metric']:.1f}% placement.")

    if any(w in q for w in ("skill", "pay", "premium", "critical", "demand")):
        prem = ranking(conn, "skill_salary_premium", module)
        if prem:
            evidence["skill_premium"] = prem[:5]
            cite(f"Top skill premium: {prem[0]['entity']} at {prem[0]['metric']:+.1f}% over median.")

    if any(w in q for w in ("channel", "source", "hire", "recruit")):
        ch = ranking(conn, "channel_effectiveness", module)
        if ch:
            evidence["channels"] = ch
            cite(f"{ch[0]['entity']} is the highest-yield channel at {ch[0]['metric']:.1f}%.")

    if any(w in q for w in ("ctc", "salary", "pay", "forecast", "next", "future", "project")):
        fc = forecast(conn, "median_ctc_cycle", module)
        proj = next((p for p in fc if p["is_forecast"]), None)
        if proj:
            evidence["ctc_forecast_next"] = proj["yhat"]
            cite(f"Median CTC projected to ₹{proj['yhat']:.1f} LPA next cycle.")

    if any(w in q for w in ("sector", "revenue", "decline", "fall", "fell", "drop", "it services")):
        recs = recommendations(conn, module)
        it = next((r for r in recs if "IT Services" in r["title"] or "IT Services" in r["observation"]), None)
        if it:
            evidence["declining_sector"] = {"title": it["title"], "observation": it["observation"]}
            cite(it["observation"])

    if any(w in q for w in ("invest", "budget", "allocate", "should we", "priorit")):
        recs = recommendations(conn, module)
        evidence["recommendations"] = [
            {"title": r["title"], "action": r["recommended_action"], "priority": r["priority"]}
            for r in recs[:4]
        ]

    return {"evidence": evidence, "citations": citations}


def answer(conn, question: str, persona: str = "executive", module=MODULE) -> dict:
    retrieved = retrieve(conn, question, module)
    evidence, citations = retrieved["evidence"], retrieved["citations"]

    ai_text = None
    if ai.is_configured():
        import json
        prompt = (
            f'Question: "{question}"\n\n'
            "Answer the question using ONLY the evidence below. Cite the figures you use. "
            "If the evidence does not answer it, say so plainly.\n\nEVIDENCE:\n"
            + json.dumps(evidence, indent=2, default=str)
        )
        ai_text = ai.generate(ai.PERSONAS.get(persona, ai.PERSONAS["executive"]), prompt)

    if ai_text:
        return {
            "question": question, "answer": ai_text, "citations": citations,
            "evidence": evidence,
            "provenance": {"engine": "ai-enhanced", "provider": ai.provider_label(), "persona": persona},
        }

    # Deterministic answer: stitch the retrieved citations into prose.
    body = " ".join(citations) if citations else "No structured evidence matched this question."
    return {
        "question": question,
        "answer": f"Based on the {module} analytics: {body}",
        "citations": citations, "evidence": evidence,
        "provenance": {"engine": "deterministic", "provider": ai.provider_label(), "persona": persona},
    }
